clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -375,6 +382,19 @@ public void wakeup() {
         this.client.wakeup();
     }
 
+    /* metadata becomes "stale" for batch expiry purpose when the time since the last successful update exceeds
+     * the metadataStaleMs value. This value must be greater than the metadata.max.age and some delta to account
+     * for a few retries and transient network disconnections. A small number of retries (3) are chosen because 
```

Comment:
Discussed offline. This can instead be the producer's retry count. If the retry count is zero, then we will have to allow for at least one metadata request past its max age. So the staleness threshold will be: `retryCount * (backoff + requestTimeout) + maxAge`